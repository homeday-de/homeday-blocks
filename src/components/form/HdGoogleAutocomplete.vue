<template>
  <TextFieldBase
    v-bind="$attrs"
    :name="name"
    :error="error"
    :helper="helper"
    :active="isActive"
    :filled="isFilled"
    :valid="isValid"
    :disabled="disabled"
    @clear-click="clearInput"
    @status-click="focusInput"
  >
    <input
      ref="input"
      :value="value"
      :id="name"
      :name="name"
      :placeholder="isActive && placeholder !== undefined ? placeholder : ''"
      :required="required"
      :autofocus="autofocus"
      :disabled="disabled"
      autocomplete="off"
      class="field__input"
      type="text"
      @input="handleInput"
      @focus="handleFocus"
      @blur="handleBlur"
    >
    <template #input-right>
      <slot name="input-right" />
    </template>
  </TextFieldBase>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import { getGoogleAPI } from 'homeday-blocks/src/services/googleAPI';
import TextFieldBase from './TextFieldBase.vue';

export default {
  name: 'HdGoogleAutocomplete',
  components: {
    TextFieldBase,
  },
  props: {
    apiKey: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    value: {
      type: String,
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    customRules: {
      type: Array,
      default: () => [],
      validator: rulesProvided => rulesProvided.every(
        ({ validate, errorMessage }) => typeof validate === 'function' && typeof errorMessage === 'string',
      ),
    },
  },
  data() {
    return {
      autocompleteInstance: null,
      geocoderInstance: null,
      lastLocation: null,
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isFilled() {
      return this.value !== null && this.value !== undefined && this.value !== '';
    },
  },
  watch: {
    value() {
      this.validate();
    },
  },
  created() {
    this.initAutocomplete();
  },
  methods: {
    async initAutocomplete() {
      await getGoogleAPI(this.apiKey);

      const autocompleteOptions = {
        componentRestrictions: { country: 'de' },
        types: ['(regions)'],
      };

      if (!this.$refs.input) {
        return;
      }

      this.autocompleteInstance = new window.google.maps.places.Autocomplete(
        this.$refs.input,
        autocompleteOptions,
      );

      window.google.maps.event.addListener(this.autocompleteInstance, 'place_changed', () => {
        const place = this.autocompleteInstance.getPlace();

        this.$emit('input', place.name);

        if (!place.geometry) {
          return;
        }

        // If the selected the whole Germany, we ignore the lat and lng
        if (place.address_components[0].short_name === 'DE') {
          this.emitLocation({
            lat: null,
            lng: null,
            name: place.name,
          });
          return;
        }

        this.emitLocation({
          lat: place.geometry.location.lat(),
          lng: place.geometry.location.lng(),
          name: place.name,
        });
      });
    },
    emitLocation({ lat = null, lng = null, name = '' }) {
      this.lastLocation = {
        lat,
        lng,
        name,
      };

      this.$emit('locationChanged', {
        lat,
        lng,
        name,
      });
    },
    async getLocation() {
      if (!this.value) {
        return {
          lat: null,
          lng: null,
          name: '',
        };
      }

      if (this.lastLocation && this.lastLocation.name === this.value) {
        return this.lastLocation;
      }

      if (!this.geocoderInstance) {
        this.geocoderInstance = new window.google.maps.Geocoder();
      }

      return new Promise((resolve) => {
        this.geocoderInstance.geocode({ address: this.value }, (results, status) => {
          if (status !== window.google.maps.GeocoderStatus.OK) {
            resolve({
              lat: null,
              lng: null,
              name: this.value,
            });
          }

          this.lastLocation = {
            lat: results[0].geometry.location.lat(),
            lng: results[0].geometry.location.lng(),
            name: this.value,
          };

          resolve(this.lastLocation);
        });
      });
    },
    clearInput() {
      this.focusInput();
      this.hideError();
      this.$emit('input', '');
    },
    focusInput() {
      this.$refs.input.focus();
    },
    handleFocus() {
      this.isActive = true;
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
    },
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    hideHelper() {
      this.helper = null;
    },
    validate() {
      if (this.required && !this.isFilled) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else if (this.customRules.length && this.isFilled) {
        this.validateCustomRules();
      } else {
        this.hideError();
      }
      return !this.error;
    },
    validateCustomRules() {
      const firstFailingRule = this.customRules.find(({ validate }) => !validate(this.value));
      if (firstFailingRule) {
        this.showError(firstFailingRule.errorMessage);
      } else {
        this.hideError();
      }
    },
  },
};
</script>
