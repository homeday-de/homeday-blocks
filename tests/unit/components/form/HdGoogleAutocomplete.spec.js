import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdGoogleAutocomplete from '@/components/form/HdGoogleAutocomplete.vue';

const TEST_VALUE = 'Berlin';
const BERLIN_LOCATION = {
  lat: 52.520008,
  lng: 13.404954,
  name: 'Berlin',
};
const DEFAULT_LOCATION = {
  lat: null,
  lng: null,
  name: '',
};

const wrapperBuilder = wrapperFactoryBuilder(HdGoogleAutocomplete, {
  props: {
    name: 'Test autocomplete',
  },
});

describe('HdGoogleAutocomplete', () => {
  it('The component is rendered', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Emits input event', async () => {
    const wrapper = wrapperBuilder();

    wrapper.find('input').setValue(TEST_VALUE);
    expect(wrapper.emitted('input')).toBeTruthy();

    const payload = wrapper.emitted('input')[0][0];

    expect(payload).toBe(TEST_VALUE);
  });

  it('Emits locationChanged event', () => {
    const wrapper = wrapperBuilder();

    wrapper.vm.emitLocation(BERLIN_LOCATION);
    expect(wrapper.emitted('locationChanged')).toBeTruthy();
    // And that expected payload is passed with the event
    const payload = wrapper.emitted('locationChanged')[0][0];
    expect(payload).toEqual(BERLIN_LOCATION);
  });

  it('Supports getLocation method with a specific signature', async () => {
    const wrapper = wrapperBuilder();
    const location = await wrapper.vm.getLocation();

    expect(location).toEqual(DEFAULT_LOCATION);
  });

  it('Supports clearing of the input', () => {
    const wrapper = wrapperBuilder();

    wrapper.get('input').setValue(TEST_VALUE);
    wrapper.vm.clearInput();
    expect(wrapper.emitted('input')).toBeTruthy();

    const payload = wrapper.emitted('input')[1][0];

    expect(payload).toBe('');
  });

  it('Supports disabling of the input', () => {
    const wrapper = wrapperBuilder({
      props: {
        disabled: true,
      },
    });

    expect(wrapper.find('input').attributes().disabled).toBe('disabled');
  });


  describe('Supports custom rules with error message', () => {
    const atLeastTenCharsErrorMessage = 'should be at least ten characters long';
    const atLeastTenChars = { validate: value => value.length >= 10, errorMessage: atLeastTenCharsErrorMessage };

    const startsWithSpaceErrorMessage = 'should start with empty space';
    const startsWithSpace = { validate: value => value[0] === ' ', errorMessage: startsWithSpaceErrorMessage };

    it('Simple case', () => {
      // invalid
      const wrapper = wrapperBuilder({
        props: {
          customRules: [atLeastTenChars],
          value: 'too short',
        },
      });

      wrapper.vm.validate();

      expect(wrapper.vm.isValid).toBe(false);
      expect(wrapper.vm.error).toBe(atLeastTenCharsErrorMessage);

      // valid
      wrapper.setProps({
        customRules: [atLeastTenChars],
        value: 'this one is long enough',
      });

      wrapper.vm.validate();

      expect(wrapper.vm.isValid).toBe(true);
      expect(wrapper.vm.error).toBeFalsy();
    });

    it('Multiple custom validators: first one has highest importance', () => {
      // both invalid
      const wrapper = wrapperBuilder({
        props: {
          customRules: [startsWithSpace, atLeastTenChars],
          value: 'short',
        },
      });

      wrapper.vm.validate();

      expect(wrapper.vm.isValid).toBe(false);
      expect(wrapper.vm.error).toBe(startsWithSpaceErrorMessage);

      // first one valid, second one invalid
      wrapper.setProps({
        customRules: [startsWithSpace, atLeastTenChars],
        value: ' short',
      });

      wrapper.vm.validate();

      expect(wrapper.vm.isValid).toBe(false);
      expect(wrapper.vm.error).toBe(atLeastTenCharsErrorMessage);
    });
  });
});
