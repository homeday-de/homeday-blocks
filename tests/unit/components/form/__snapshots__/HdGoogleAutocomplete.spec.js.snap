// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdGoogleAutocomplete The component is rendered 1`] = `
<div class="field text-field field--unlabelled">
  <!---->
  <div class="field__body">
    <div class="field__main"><input id="Test autocomplete" name="Test autocomplete" placeholder="" autocomplete="off" type="text" class="field__input">
      <!---->
      <div class="field__input-right"></div>
      <div class="field__border"></div>
    </div>
    <p class="field__helper">‚Äç</p>
  </div>
</div>
`;
